File: src/main/java/com/yupi/airouter/plugin/impl/ImageRecognitionPlugin.java
/**
 * 图片识别插件
 * 识别图片内容，返回图片描述
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
package com.yupi.airouter.plugin.impl;

import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONObject;
import cn.hutool.json.JSONUtil;
import com.mybatisflex.core.query.QueryWrapper;
import com.yupi.airouter.model.entity.ModelProvider;
import com.yupi.airouter.plugin.Plugin;
import com.yupi.airouter.plugin.PluginContext;
import com.yupi.airouter.plugin.PluginResult;
import com.yupi.airouter.service.ModelProviderService;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

@Slf4j
@Component
public class ImageRecognitionPlugin implements Plugin {

    private static final String PLUGIN_KEY = "image_recognition";
    private static final String PLUGIN_NAME = "图片识别";
    private static final String DESCRIPTION = "识别图片内容，返回图片描述";

    /**
     * 使用的视觉模型
     */
    private String model = "qwen-vl-plus";

    /**
     * 最大图片大小（字节）
     */
    private int maxImageSize = 4 * 1024 * 1024; // 4MB

    /**
     * 默认提示词
     */
    private String defaultPrompt = "请描述这张图片的内容，包括主要对象、场景、颜色、文字等信息。";

    @Resource
    private ModelProviderService modelProviderService;

    /**
     * HTTP 客户端（支持重定向）
     */
    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build();

    @Override
    public String getPluginKey() {
        return PLUGIN_KEY;
    }

    @Override
    public String getPluginName() {
        return PLUGIN_NAME;
    }

    @Override
    public String getDescription() {
        return DESCRIPTION;
    }

    @Override
    public void init(String config) {
        if (StrUtil.isNotBlank(config)) {
            try {
                JSONObject configJson = JSONUtil.parseObj(config);
                this.model = configJson.getStr("model", "qwen-vl-plus");
                this.maxImageSize = configJson.getInt("maxImageSize", 4 * 1024 * 1024);
                this.defaultPrompt = configJson.getStr("defaultPrompt", defaultPrompt);
                log.info("图片识别插件初始化: model={}, maxImageSize={}", model, maxImageSize);
            } catch (Exception e) {
                log.warn("解析图片识别插件配置失败，使用默认配置", e);
            }
        }
    }

    @Override
    public boolean supports(PluginContext context) {
        // 需要有图片 URL 或文件字节数组
        return StrUtil.isNotBlank(context.getFileUrl()) 
                || (context.getFileBytes() != null && context.getFileBytes().length > 0);
    }

    @Override
    public PluginResult execute(PluginContext context) {
        log.info("执行图片识别");

        try {
            // 获取图片数据
            byte[] imageBytes;
            String mimeType = "image/jpeg";

            // 获取图片数据
            if (context.getFileBytes() != null && context.getFileBytes().length > 0) {
                // 直接使用上传的文件字节数组
                imageBytes = context.getFileBytes();
                if (StrUtil.isNotBlank(context.getFileType())) {
                    mimeType = context.getFileType();
                }
                log.info("使用上传的图片，大小: {} bytes", imageBytes.length);
            } else if (StrUtil.isNotBlank(context.getFileUrl())) {
                // 从 URL 下载
                imageBytes = downloadImage(context.getFileUrl());
                mimeType = guessMimeType(context.getFileUrl());
                log.info("从URL下载图片，大小: {} bytes", imageBytes.length);
            } else {
                return PluginResult.fail("请提供图片文件或图片URL");
            }

            // 检查图片大小
            if (imageBytes.length > maxImageSize) {
                return PluginResult.fail("图片大小超过限制（最大 " + (maxImageSize / 1024 / 1024) + "MB）");
            }

            // 获取用户问题或使用默认提示词
            String prompt = StrUtil.isNotBlank(context.getInput()) ? context.getInput() : defaultPrompt;

            // 调用视觉模型
            String result = recognizeImage(imageBytes, mimeType, prompt);

            // 构建结果
            StringBuilder contentBuilder = new StringBuilder();
            contentBuilder.append("## 图片识别结果\n\n");
            if (StrUtil.isNotBlank(context.getInput())) {
                contentBuilder.append("用户问题: ").append(context.getInput()).append("\n\n");
            }
            contentBuilder.append(result);

            // 构建额外数据
            Map<String, Object> data = new HashMap<>();
            data.put("model", model);
            data.put("imageSize", imageBytes.length);

            return PluginResult.success(contentBuilder.toString(), data);
        } catch (Exception e) {
            log.error("图片识别失败", e);
            return PluginResult.fail("图片识别失败: " + e.getMessage());
        }
    }

    /**
     * 下载图片
     */
    private byte[] downloadImage(String url) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(30))
                .GET()
                .build();

        HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());

        if (response.statusCode() != 200) {
            throw new RuntimeException("下载图片失败，状态码: " + response.statusCode());
        }

        return response.body();
    }

    /**
     * 猜测图片 MIME 类型
     */
    private String guessMimeType(String url) {
        url = url.toLowerCase();
        if (url.contains(".png")) {
            return "image/png";
        } else if (url.contains(".gif")) {
            return "image/gif";
        } else if (url.contains(".webp")) {
            return "image/webp";
        }
        return "image/jpeg";
    }

    /**
     * 调用视觉模型识别图片
     * 使用 HTTP API 直接调用通义千问视觉模型
     */
    private String recognizeImage(byte[] imageBytes, String mimeType, String prompt) {
        // 获取通义千问提供者
        ModelProvider provider = modelProviderService.getOne(
                QueryWrapper.create().eq("providerName", "qwen"));
        if (provider == null) {
            throw new RuntimeException("未找到通义千问提供者配置");
        }

        try {
            // 将图片转换为 Base64
            String base64Image = Base64.getEncoder().encodeToString(imageBytes);
            String imageUrl = "data:" + mimeType + ";base64," + base64Image;

            // 构建请求体（通义千问视觉模型 API 格式）
            JSONObject requestBody = new JSONObject();
            requestBody.set("model", model);
            
            // 构建消息
            JSONObject userMessage = new JSONObject();
            userMessage.set("role", "user");
            
            // 构建多模态内容
            cn.hutool.json.JSONArray contentArray = new cn.hutool.json.JSONArray();
            
            // 添加图片
            JSONObject imageContent = new JSONObject();
            imageContent.set("type", "image_url");
            JSONObject imageUrlObj = new JSONObject();
            imageUrlObj.set("url", imageUrl);
            imageContent.set("image_url", imageUrlObj);
            contentArray.add(imageContent);
            
            // 添加文本
            JSONObject textContent = new JSONObject();
            textContent.set("type", "text");
            textContent.set("text", prompt);
            contentArray.add(textContent);
            
            userMessage.set("content", contentArray);
            
            cn.hutool.json.JSONArray messagesArray = new cn.hutool.json.JSONArray();
            messagesArray.add(userMessage);
            requestBody.set("messages", messagesArray);

            // 发送请求
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(provider.getBaseUrl() + "/v1/chat/completions"))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + provider.getApiKey())
                    .timeout(Duration.ofSeconds(60))
                    .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                log.error("视觉模型调用失败，状态码: {}, 响应: {}", response.statusCode(), response.body());
                throw new RuntimeException("视觉模型调用失败，状态码: " + response.statusCode());
            }

            // 解析响应
            JSONObject responseJson = JSONUtil.parseObj(response.body());
            cn.hutool.json.JSONArray choices = responseJson.getJSONArray("choices");
            if (choices != null && !choices.isEmpty()) {
                JSONObject choice = choices.getJSONObject(0);
                JSONObject message = choice.getJSONObject("message");
                if (message != null) {
                    return message.getStr("content", "无法识别图片内容");
                }
            }

            throw new RuntimeException("模型返回结果格式错误");
        } catch (Exception e) {
            log.error("调用视觉模型失败", e);
            throw new RuntimeException("调用视觉模型失败: " + e.getMessage());
        }
    }
}

File: python-backend/app/plugin/impl/image_recognition_plugin.py
"""Image recognition plugin."""

from __future__ import annotations

import base64
import json

import httpx
from sqlalchemy import select

from app.db.session import session_maker
from app.models.model_provider import ModelProvider
from app.plugin.plugin import Plugin
from app.plugin.plugin_context import PluginContext
from app.plugin.plugin_result import PluginResult


class ImageRecognitionPlugin(Plugin):
    PLUGIN_KEY = "image_recognition"
    PLUGIN_NAME = "图片识别"
    DESCRIPTION = "识别图片内容，返回图片描述"

    def __init__(self) -> None:
        self.model = "qwen-vl-plus"
        self.max_image_size = 4 * 1024 * 1024
        self.default_prompt = "请描述这张图片的内容，包括主要对象、场景、颜色、文字等信息。"

    def get_plugin_key(self) -> str:
        return self.PLUGIN_KEY

    def get_plugin_name(self) -> str:
        return self.PLUGIN_NAME

    def get_description(self) -> str:
        return self.DESCRIPTION

    def init(self, config: str | None) -> None:
        if not config:
            return
        try:
            data = json.loads(config)
            self.model = str(data.get("model", self.model))
            self.max_image_size = int(data.get("maxImageSize", self.max_image_size))
            self.default_prompt = str(data.get("defaultPrompt", self.default_prompt))
        except Exception:
            return

    def supports(self, context: PluginContext) -> bool:
        return bool(context.file_url) or bool(context.file_bytes)

    async def execute(self, context: PluginContext) -> PluginResult:
        try:
            image_bytes, mime_type = await self._resolve_image(context)
            if len(image_bytes) > self.max_image_size:
                return PluginResult.fail(f"图片大小超过限制（最大 {self.max_image_size // 1024 // 1024}MB）")
            prompt = (context.input or "").strip() or self.default_prompt
            content = await self._recognize(image_bytes, mime_type, prompt)
            text = content
            if context.input:
                text = f"用户问题: {context.input}\n\n{content}"
            return PluginResult.success_result(
                text,
                {"model": self.model, "imageSize": len(image_bytes)},
            )
        except Exception as exc:
            return PluginResult.fail(f"图片识别失败: {exc}")

    async def _resolve_image(self, context: PluginContext) -> tuple[bytes, str]:
        if context.file_bytes:
            return context.file_bytes, context.file_type or "image/jpeg"
        if not context.file_url:
            raise RuntimeError("请提供图片文件或图片URL")
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(context.file_url)
        if resp.status_code != 200:
            raise RuntimeError(f"下载图片失败，状态码: {resp.status_code}")
        mime = self._guess_mime(context.file_url)
        return resp.content, mime

    async def _recognize(self, image_bytes: bytes, mime_type: str, prompt: str) -> str:
        provider = await self._get_qwen_provider()
        if provider is None:
            raise RuntimeError("未找到通义千问提供者配置")
        image_url = f"data:{mime_type};base64,{base64.b64encode(image_bytes).decode('utf-8')}"
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {"type": "image_url", "image_url": {"url": image_url}},
                        {"type": "text", "text": prompt},
                    ],
                }
            ],
        }
        url = f"{provider.base_url.rstrip('/')}/v1/chat/completions"
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {provider.api_key}",
        }
        async with httpx.AsyncClient(timeout=60) as client:
            resp = await client.post(url, headers=headers, json=payload)
        if resp.status_code != 200:
            raise RuntimeError(f"视觉模型调用失败，状态码: {resp.status_code}")
        data = resp.json()
        choices = data.get("choices") or []
        if not choices:
            raise RuntimeError("模型返回结果格式错误")
        message = choices[0].get("message") or {}
        return str(message.get("content") or "无法识别图片内容")

    @staticmethod
    async def _get_qwen_provider() -> ModelProvider | None:
        async with session_maker() as db:
            stmt = select(ModelProvider).where(
                ModelProvider.provider_name == "qwen",
                ModelProvider.is_delete == 0,
            )
            return await db.scalar(stmt)

    @staticmethod
    def _guess_mime(url: str) -> str:
        lowered = url.lower()
        if ".png" in lowered:
            return "image/png"
        if ".gif" in lowered:
            return "image/gif"
        if ".webp" in lowered:
            return "image/webp"
        return "image/jpeg"
